### PR TITLE
Define a type for onChange func instead of func pointer

### PR DESCRIPTION
**Pull request by bogdandrutu**
Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: config/configmapprovider/provider.go [hunk 1]
```diff
@@ -62,6 +62,8 @@ type Provider interface {
 	Shutdown(ctx context.Context) error
 }
 
+type Watcher func(*ChangeEvent)
```
Comment (bogdandrutu) [inline config/configmapprovider/provider.go:65]: Other name suggestions are also appreciate.

Comment (tigrannajaryan) [reply config/configmapprovider/provider.go:65]: `WatcherFunc` perhaps?

**Review by tigrannajaryan — APPROVED** [review]

File: config/configmapprovider/provider.go (hunk 1 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go:65]: Good suggestion, will do.

File: config/configmapprovider/provider.go [hunk 2]
```diff
@@ -40,15 +40,15 @@ type Provider interface {
 	// contains the value to be injected in the configuration and the corresponding watcher that
 	// will be used to monitor for updates of the retrieved value.
 	//
-	// onChange callback is called when the config changes. onChange may be called from
+	// watcher callback is called when the config changes. onChange may be called from
 	// a different go routine. After onChange is called Retrieved.Get should be called
```
Comment (codeboten) [inline config/configmapprovider/provider.go:44]:
```suggestion
	// a different go routine. After watcher is called Retrieved.Get should be called
```

File: config/configmapprovider/provider.go [hunk 3]
```diff
@@ -40,15 +40,15 @@ type Provider interface {
...
 	// to get the new config. See description of Retrieved for more details.
 	// onChange may be nil, which indicates that the caller is not interested in
```
Comment (codeboten) [inline config/configmapprovider/provider.go:46]:
```suggestion
	// watcher may be nil, which indicates that the caller is not interested in
```

File: service/config_provider.go [hunk 4]
```diff
@@ -145,7 +145,7 @@ func (cm *configProvider) Watch() <-chan error {
 }
 
 func (cm *configProvider) onChange(event *configmapprovider.ChangeEvent) {
-	// TODO: Remove check for configsource.ErrSessionClosed when providers updated to not call onChange when closed.
+	// TODO: Remove check for configsource.ErrSessionClosed when providers updated to not call watcher when closed.
```
Comment (codeboten) [inline service/config_provider.go:148]: should this have been changed? the method name here is still onChange

**Review by codeboten — COMMENTED** [review]
Just confirming this is a change to improve the usability of the interface? I didn't see an issue associated or a description for the change.

**Comment by bogdandrutu** [comment]
@codeboten yes, this PR is just to offer a more intuitive API, essentially instead of relying on the name of the variable to suggest what the function does, we define a type with a good name.

File: service/config_provider.go (hunk 4 above)
Comment (bogdandrutu) [reply service/config_provider.go]: GoLand decided to do this.

File: config/configmapprovider/provider.go (hunk 3 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go]: Done.

File: config/configmapprovider/provider.go (hunk 2 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go]: Done.

File: config/configmapprovider/provider.go (hunk 1 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go]: Done.

**Review by Aneurysm9 — APPROVED** [review]

**Pull request merged**